modules/platform/nuxeo-drive-server/nuxeo-drive-mongodb/src/main/java/org/nuxeo/drive/mongodb/MongoDBAuditChangeFinder.java

```diff
@@ -0,0 +1,232 @@
+                    return 0;
+                }
+            }
+            log.debug("Found no audit log entries, returning -1");
+            return -1;
+        }
+        return entries.get(0).getId();
+    }
+
+    @Override
+    protected List<LogEntry> queryAuditEntries(CoreSession session, SynchronizationRoots activeRoots,
+            Set<String> collectionSyncRootMemberIds, long lowerBound, long upperBound, int limit) {
+        MongoDBAuditBackend auditService = (MongoDBAuditBackend) Framework.getService(AuditReader.class);
+        MongoCollection<Document> auditCollection = auditService.getAuditCollection();
+        StringBuilder queryBuilder = new StringBuilder();
+        Map<String, Object> params = new HashMap<>();
+        params.put("repositoryId", session.getRepositoryName());
+        params.put("lowerBound", lowerBound);
+        params.put("upperBound", upperBound);
+
+        if(queryStart == null) {
+            try (InputStream is = getClass().getClassLoader().getResourceAsStream("OSGI-INF/queryStart.json")) {
+                queryStart = IOUtils.toString(is, "UTF-8");
+            } catch (IOException e) {
+                log.error("couldn't read query start template", e);
+            }
+        }
+        queryBuilder.append(queryStart);
+        // first conditional filter
+        if (!activeRoots.getPaths().isEmpty()) {
+            queryBuilder.append(",\n");
+            // We open an $or array here in case the next conditional filter is needed as it's this or that
+            queryBuilder.append("{ \"$or\" : [");
+            queryBuilder.append(getCurrentRootFilteringClause(activeRoots.getPaths(), params) + "\n");
+        }
+        // second conditional filter
+        if (collectionSyncRootMemberIds != null && !collectionSyncRootMemberIds.isEmpty()) {
+            queryBuilder.append(",\n");
+            queryBuilder.append(getCollectionSyncRootFilteringClause(collectionSyncRootMemberIds, params) + "\n");
+        }
+        // if the first conditional filter was set, then an $or array has been opened and needs to be closed
+        if (!activeRoots.getPaths().isEmpty()) {
+            queryBuilder.append("]}\n");
+        }
+        if(queryEnd == null) {
+            try (InputStream is = getClass().getClassLoader().getResourceAsStream("OSGI-INF/queryEnd.json")) {
+                queryEnd = IOUtils.toString(is, "UTF-8");
+            } catch (IOException e) {
+                log.error("couldn't read query end template", e);
+            }
+        }
+        queryBuilder.append(queryEnd);
+
+        Bson filter = auditService.buildFilter(queryBuilder.toString(), params);
+        var sorts = List.of(Sorts.ascending("repositoryId"), Sorts.descending("eventDate"));
+        var order = Sorts.orderBy(sorts);
+        FindIterable<Document> iterable = auditCollection.find(filter).sort(order);
+        List<LogEntry> entries = StreamSupport.stream(iterable.spliterator(), false)
```

Comment:
`FindIterable#spliterator` is backed by the default method in `Iterable` which leverages `Iterable#iterator`, or in this case `MongoIterableImpl#iterator` which returns a `MongoCursor` which needs to be closed. You have an example of what to do in [MongoDBConnection](https://github.com/nuxeo/nuxeo/blob/927293a3f326fbcc8d8662ecee35ff6a2d5c8601/modules/core/nuxeo-core-storage-mongodb/src/main/java/org/nuxeo/ecm/core/storage/mongodb/MongoDBConnection.java#L698).

Otherwise, you could use the MongoDB API there, something like:

```java
List<LogEntry> entries = auditCollection.find(filter).sort(order).map(MongoDBAuditEntryReader::read).into(new ArrayList<>());
```